marshal.go

```diff
@@ -0,0 +1,279 @@
+	paramsByName := make(map[string]string)
+	for _, param := range p.PathVar {
+		paramsByName[param.Key] = param.Value
+	}
+
+	offset := 0
+	hasParams := false
+	for i := 0; i < len(path); i++ {
+		c := path[i]
+		if c != ':' && c != '*' {
+			continue
+		}
+		hasParams = true
+
+		end := i + 1
+		for end < len(path) && path[end] != ':' && path[end] != '/' {
+			end++
+		}
+
+		if c == '*' && end != len(path) {
+			return errgo.New("placeholders starting with * are only allowed at the end")
+		}
+
+		if end-i < 2 {
+			return errgo.New("request wildcards must be named with a non-empty name")
+		}
+		if i > 0 {
+			pathBuffer.WriteString(path[offset:i])
+		}
+
+		wildcard := path[i+1 : end]
+		paramValue, ok := paramsByName[wildcard]
+		if !ok {
+			return errgo.Newf("missing value for path parameter %q", wildcard)
+		}
+		pathBuffer.WriteString(paramValue)
+		offset = end
+	}
+	if !hasParams {
+		pathBuffer.WriteString(path)
+	}
+
+	p.URL.Path = pathBuffer.String()
+
+	p.URL.RawQuery = p.Form.Encode()
+
+	return nil
+}
+
+// getMarshaler returns a marshaler function suitable for marshaling
+// a field with the given tag into and http request.
+func getMarshaler(tag tag, t reflect.Type) (marshaler, error) {
+	switch {
+	case tag.source == sourceNone:
+		return marshalNop, nil
+	case tag.source == sourceBody:
+		return marshalBody, nil
+	case t == reflect.TypeOf([]string(nil)):
+		if tag.source != sourceForm {
+			return nil, errgo.New("invalid target type []string for path parameter")
+		}
+		return marshalAllField(tag.name), nil
+	case t == reflect.TypeOf(""):
+		return marshalString(tag), nil
+	case implementsTextMarshaler(t):
+		return marshalWithMarshalText(t, tag), nil
+	default:
+		return marshalWithSprint(tag), nil
+	}
+}
+
+func dereferenceIfPointer(v reflect.Value) reflect.Value {
+	if v.Kind() == reflect.Ptr {
+		if v.IsNil() {
+			return reflect.Value{}
+		}
+		return v.Elem()
+	}
+	return v
+}
+
+// marshalNop does nothing with the value.
+func marshalNop(v reflect.Value, p *Params) error {
+	return nil
+}
+
+// mashalBody marshals the specified value into the body of the http request.
+func marshalBody(v reflect.Value, p *Params) error {
+	// TODO allow body types that aren't necessarily JSON.
+	bodyValue := dereferenceIfPointer(v)
+	if bodyValue == emptyValue {
+		return nil
+	}
+
+	if p.Method != "POST" && p.Method != "PUT" {
+		return errgo.Newf("trying to marshal to body of a request with method %q", p.Method)
+	}
+
+	data, err := json.Marshal(bodyValue.Interface())
+	if err != nil {
+		return errgo.Notef(err, "cannot marshal request body")
+	}
+	p.Body = ioutil.NopCloser(bytes.NewBuffer(data))
+	return nil
+}
+
+// marshalAllField marshals a []string slice into form fields.
+func marshalAllField(name string) marshaler {
+	return func(v reflect.Value, p *Params) error {
+		value := dereferenceIfPointer(v)
+		if value == emptyValue {
+			return nil
+		}
+		values := value.Interface().([]string)
+		if p.Form == nil {
```

Comment:
I think this whole thing can just be:

    p.Form[name] = value.Interface().([]string)